specs/002-session-state-machine/research.md

```diff
@@ -0,0 +1,447 @@
+
+---
+
+## Key Architectural Decisions
+
+### 1. Shared Dictionary for Snapshots
+
+**Decision**: Use `Map[String, ByteVector]` with key prefixes for combined snapshots
+
+**Rationale**:
+- Atomic snapshot capture of both session state and user state
+- Session state uses "session/" prefix
+- User state uses "user/" prefix  
+- Avoids separate snapshot mechanisms
+- Aligns with Raft's single-snapshot model
+
+**Alternatives Considered**:
+- Separate snapshots: Rejected - complicates restoration order and atomicity
+- Nested structures: Rejected - harder to serialize/deserialize generically
+
+**Implementation**: Pass single dictionary to both state machines during snapshot/restore
+
+---
+
+### 2. State Machine Purity (No Side Effects)
+
+**Decision**: State machine is pure function: `(State, RaftAction) => UIO[(State, Response)]`
+
+**Rationale**:
+- Aligns with Raft dissertation's deterministic state machine model
+- User cannot introduce bugs via side effects
+- Testable without mocking
+- Errors modeled in response payload (FR-014: no exceptions)
+
+**Alternatives Considered**:
+- Allowing ZIO effects: Rejected - breaks determinism, hard to test
+- Exception-based errors: Rejected - violates Constitution II
+
+**Implementation**: 
+- User state machines return `(newState, response, Option[List[ServerRequest]])`
+- Session state machine wraps user SM and adds idempotency layer
+
+---
+
+### 3. Cumulative Acknowledgment for Server-Initiated Requests
+
+**Decision**: When acknowledging request ID N, remove all requests with ID ≤ N
+
+**Rationale**:
+- Aligns with TCP ACK semantics
+- Enables future protocol optimization (piggyback last ID only)
+- Works correctly with current protocol (one ack per request)
+- Simplifies implementation (single remove operation)
+- Raft ordering guarantees all lower IDs were delivered first
+
+**Alternatives Considered**:
+- Individual acknowledgment only: Rejected - requires future protocol change if optimizing
+- Out-of-order acks: Rejected - cumulative makes order irrelevant
+
+**Implementation**: `pendingRequests.filter(_.id > acknowledgedId)`
+
+---
+
+### 4. Dirty Read Optimization for Retries
+
+**Decision**: External process performs non-linearizable read before sending GetServerRequestsForRetry command
+
+**Rationale**:
+- Avoids unnecessary Raft log entries when no retries needed
+- Safe because command response is authoritative (dirty read just hint)
+- Process discards dirty data after decision to skip/send command
+- Worst case: one missed retry cycle if stale read
+
+**Alternatives Considered**:
+- Always send command: Rejected - creates unnecessary log growth during idle periods
+- Maintain external counter: Rejected - adds complexity, can get out of sync
+
+**Implementation**:
+```scala
+// Dirty read (no ZIO effect needed - direct state access)
+val mayNeedRetry = stateMachine.hasPendingRequests()
```

Comment:
hasPendingRequests is a actually important method, let's add it as public method to SessionStateMachine in the data-model document.